### PR TITLE
feat(chisel): Add edit command

### PR DESCRIPTION
**Pull request by fvictorio**
## Motivation

To make a change in the current code of the chisel session, you have to clear it and start over. It would be nice to be able to edit it with your `$EDITOR`.

## Solution

Add an `!edit` command that opens the (run) code of the current session with the program set in the `$EDITOR` variable (or vim, if no editor is set).

---

This is more of a proof of concept and the code is... played by ear, let's say. If the feature seems useful and someone who actually knows Rust wants to implement it from scratch, feel free to close this PR!

Also: I don't like that the code inside the function is the one being edited. I would prefer for the whole "file" to be edited. But I don't thinks that's easy to do right now? You would need to parse the file to reconstruct the session, which I'm not sure it's worth the effort.

### CODE REVIEW

**Review by mattsse — APPROVED** [review]
useful,

wdyt @clabby

**Review by clabby — COMMENTED** [review]
Love the idea, this is pretty useful imo. Agreed that not being able to edit the whole file is sub-optimal, we may want to rework the way that the `SessionSource` stores separately scoped code at some point. For now, a good workaround might be to add another argument to this command that designates whether the `global`, `top-level`, or `run` scope is being edited.

File: chisel/src/dispatcher.rs [hunk 1]
```diff
@@ -610,6 +610,71 @@ impl ChiselDisptacher {
...
+                    // if the editor exited successfully, try to compile the new code
+                    let edited_code = std::fs::read_to_string(file_path).unwrap();
+                    let mut new_session_source = session_source.clone();
+                    new_session_source.drain_run();
+                    new_session_source.with_run_code(&edited_code);
+                    match new_session_source.execute().await {
+                        Ok((_, res)) => {
+                            let failed = !res.success;
+                            if failed {
+                                DispatchResult::Failure(None)
```
Comment (clabby) [inline chisel/src/dispatcher.rs:663]: nit- use `DispatchResult::CommandFailed` here, and give some feedback as to why the edit failed.

File: chisel/src/dispatcher.rs [hunk 2]
```diff
@@ -610,6 +610,71 @@ impl ChiselDisptacher {
...
+                        Err(_) => {
+                            return DispatchResult::CommandFailed(
+                                "Editor exited without a status code".to_string(),
+                            )
+                        }
+                    }
+
+                    // if the editor exited successfully, try to compile the new code
+                    let edited_code = std::fs::read_to_string(file_path).unwrap();
+                    let mut new_session_source = session_source.clone();
```
Comment (clabby) [inline chisel/src/dispatcher.rs:656]: We may want to expose the `SessionSource`'s `shallow_clone` function for this, but not necessary.

File: chisel/src/dispatcher.rs [hunk 3]
```diff
@@ -610,6 +610,71 @@ impl ChiselDisptacher {
...
+                    new_session_source.with_run_code(&edited_code);
+                    match new_session_source.execute().await {
+                        Ok((_, res)) => {
+                            let failed = !res.success;
+                            if failed {
+                                DispatchResult::Failure(None)
+                            } else {
+                                // the code could be compiled, save it
+                                *session_source = new_session_source;
+                                DispatchResult::CommandSuccess(None)
```
Comment (clabby) [inline chisel/src/dispatcher.rs:667]: nit- Give some feedback that the editing was successful.
```suggestion
                                DispatchResult::CommandSuccess(Some(String::from("Successfully edited `run()` function's body!")))
```
or something ^

**Review by gakonst — COMMENTED** [review]
Supportive of the change and lgtm stylistically. Just a few unwraps would be better as question marks. 


defer to vex on any further logic changes, else lmk when good to merge

**Comment by fvictorio** [comment]
@clabby thanks for the review, I applied your suggested changes.

@gakonst I considered that but I didn't see any existing `?`s in the code and, since the function it's returning an enum, I don't know how to do that without substantial changes. That being said, I removed the only `.unwrap()` I added.

**Review by mattsse — APPROVED** [review]
lgtm

**Pull request merged**